add custom model to Part 2 tutorial

yanndupis opened the pull request:
Hey @jvmancuso, 

I have updated the Part 2 tutorial to demonstrate that we can train custom models as well. 

thanks

jvmancuso commented on examples/Part 02 - Intro to Private Training with Remote Execution.ipynb:
```diff
@@ -130,7 +130,18 @@
    "cell_type": "code",
```
<p>This sounds a little too marketing-ish. I'd suggest something like:</p><p>&gt; If your model doesn't fit into the Sequential paradigm, you can use Keras's functional API, or even subclass [<code>tf.keras.Model</code>](<a href="https://www.tensorflow.org/guide/keras/custom_layers_and_models#building_models" target="_blank">https://www.tensorflow.org/guide/keras/custom_layers_and_models#building_models</a>) to create custom models.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/OpenMined/PySyft-TensorFlow/pull/33/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='21'/>

yanndupis replied:
sounds good!

yanndupis replied:
fixed

jvmancuso commented on examples/Part 02 - Intro to Private Training with Remote Execution.ipynb:
```diff
@@ -130,7 +130,18 @@
    "cell_type": "code",
```
<p>Any chance you tried wrapping <code>call</code> in a <code>tf.function</code> before sending? Would like to see if you can switch the tracing from <code>model.predict(tf.ones([1, 28, 28]))</code> to <code>model(tf.ones([1, 28, 28]))</code> by doing that.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/OpenMined/PySyft-TensorFlow/pull/33/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='22'/>

yanndupis replied:
<p>yeah I tried stuff like (with and without input_signature):</p><p> </p><p><code></code><code></p><p>class CustomModel(tf.keras.Model):</p><p> </p><p>&nbsp;&nbsp;def __init__(self, num_classes=10):</p><p>&nbsp;&nbsp;&nbsp;&nbsp;super(CustomModel, self).__init__(name='custom_model')</p><p>&nbsp;&nbsp;&nbsp;&nbsp;self.num_classes = num_classes</p><p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</p><p>&nbsp;&nbsp;&nbsp;&nbsp;self.flatten = tf.keras.layers.Flatten(input_shape=(28, 28))</p><p>&nbsp;&nbsp;&nbsp;&nbsp;self.dense_1 = tf.keras.layers.Dense(128, activation='relu')</p><p>&nbsp;&nbsp;&nbsp;&nbsp;self.dropout = tf.keras.layers.Dense(0.2)</p><p>&nbsp;&nbsp;&nbsp;&nbsp;self.dense_2 = tf.keras.layers.Dense(num_classes, activation='softmax')</p><p> </p><p>&nbsp;&nbsp;@tf.function(input_signature=[tf.TensorSpec(shape=[None, 28, 28], dtype=tf.float32)])</p><p>&nbsp;&nbsp;def call(self, inputs):</p><p>&nbsp;&nbsp;&nbsp;&nbsp;x = self.flatten(inputs)</p><p>&nbsp;&nbsp;&nbsp;&nbsp;x = self.dense_1(x)</p><p>&nbsp;&nbsp;&nbsp;&nbsp;x = self.dropout(x)</p><p>&nbsp;&nbsp;&nbsp;&nbsp;return self.dense_2(x)</p><p></code><code></code></p><p> </p><p>Even if we call 	<code>model(tf.ones([1, 28, 28]))</code>, I will get :</p><p><code>ValueError: Model &lt;__main__.CustomModel object at 0x159481588&gt; cannot be saved because the input shapes have not been set. Usually, input shapes are automatically determined from calling .fit() or .predict(). To manually set the shapes, call model._set_inputs(inputs).</code></p><p> </p><p>Maybe if we specify the input signature in [tf.keras.models.save_model](<a href="https://www.tensorflow.org/api_docs/python/tf/keras/models/save_model" target="_blank">https://www.tensorflow.org/api_docs/python/tf/keras/models/save_model</a>) that would fix it</p>

jvmancuso replied:
fair enough -- let's open an issue to track and mark it as `good first issue` with a link back here

yanndupis replied:
sounds good!

yanndupis commented:
Hey @jvmancuso, I addressed the comments. There was a bug in the custom model (instantiating a dense layer instead of dropout...)

when saving the model with a Dropout layer, I am getting:
```
ValueError: Could not find matching function to call loaded from the SavedModel. Got: Positional arguments (2 total): * Tensor("inputs:0", shape=(None, 28, 28), dtype=float32) * Tensor("training:0", shape=(), dtype=bool) Keyword arguments: {}Expected these arguments to match one of the following 4 option(s):Option 1: Positional arguments (2 total): * TensorSpec(shape=(None, 28, 28), dtype=tf.float32, name='inputs') * True Keyword arguments: {}Option 2: Positional arguments (2 total): * TensorSpec(shape=(None, 28, 28), dtype=tf.float32, name='input_1') * True Keyword arguments: {}
```

There is probably a tweak that I am missing. I removed the Dropout layer for now so we get this merged (it's a Keras pb). I will investigate this later.

jvmancuso commented:
Great, sounds good. Can we open an issue to track the `Dropout` problem so that someone else can pick it up if they want it sooner?
